### PR TITLE
ARROW-17733: [C++] Take index_width into account when filling nulls in index buffer

### PR DESCRIPTION
**Pull request by rasnjo**
Take into account index_width when offsetting by position into out_data. Otherwise we offset position bytes into the array, but we want to offset position places into the array.

### CODE REVIEW

File: cpp/src/arrow/array/concatenate_test.cc [hunk 1]
```diff
@@ -539,4 +539,13 @@ TEST_F(ConcatenateTest, OffsetOverflow) {
   ASSERT_RAISES(Invalid, Concatenate({fake_long, fake_long}).status());
 }
 
+TEST_F(ConcatenateTest, DictionaryConcatenateWithEmptyUint16) {
+  auto dict_type = dictionary(uint16(), utf8());
+  auto dict_one = DictArrayFromJSON(dict_type, "[]", "[]");
+  auto dict_two = DictArrayFromJSON(dict_type, "[0, 1, null]", "[\"A0\", \"A1\"]");
+  ASSERT_OK_AND_ASSIGN(auto concat_actual, Concatenate({dict_one, dict_two}));
```
Comment (lidavidm) [inline cpp/src/arrow/array/concatenate_test.cc:546]: nit: it might be worth concatenating with a run of a few nulls?

**Review by lidavidm — APPROVED** [review]
Good catch, and thank you for fixing this!

File: cpp/src/arrow/array/concatenate_test.cc (hunk 1 above)
Comment (rasnjo) [reply cpp/src/arrow/array/concatenate_test.cc:548]: good suggestion, makes sense. Done in commit: 204981dc56a71b628286a9e9264ad634a3c90d8d

File: cpp/src/arrow/array/concatenate_test.cc [hunk 2]
```diff
@@ -539,4 +539,14 @@ TEST_F(ConcatenateTest, OffsetOverflow) {
   ASSERT_RAISES(Invalid, Concatenate({fake_long, fake_long}).status());
 }
 
+TEST_F(ConcatenateTest, DictionaryConcatenateWithEmptyUint16) {
+  auto dict_type = dictionary(uint16(), utf8());
```
[the remark this answers is not in the document]
Comment (rasnjo) [reply cpp/src/arrow/array/concatenate_test.cc:544]: Yes, done now (commit: 8d02222c2efd71987f33448bbc9a5546100ba360)

**Comment by lidavidm** [comment]
Alright, the test failures look to be unrelated to what's going on here/are tracked elsewhere, so this should be good. Thanks!

**Pull request merged**